ray-operator/controllers/raycluster_controller_fake_test.go

```diff
@@ -0,0 +1,426 @@
+	assert.Nil(t, err, "Fail to create requirement")
+	groupNameReqValue := []string{groupNameStr}
+	groupNameReq, err := labels.NewRequirement(
+		common.RayNodeGroupLabelKey,
+		selection.Equals,
+		groupNameReqValue)
+	assert.Nil(t, err, "Fail to create requirement")
+
+	workerSelector = labels.NewSelector().Add(*instanceReq).Add(*groupNameReq)
+}
+
+func tearDown(t *testing.T) {
+	PrioritizeWorkersToDelete = false
+}
+
+func TestReconcile_RemoveWorkersToDelete_OK(t *testing.T) {
+	setupTest(t)
+	defer tearDown(t)
+
+	fakeClient := clientFake.NewClientBuilder().WithRuntimeObjects(testPods...).Build()
+
+	podList := corev1.PodList{}
+	err := fakeClient.List(context.Background(), &podList, client.InNamespace(namespaceStr))
+
+	assert.Nil(t, err, "Fail to get pod list")
+	assert.Equal(t, len(testPods), len(podList.Items), "Init pod list len is wrong")
+
+	testRayClusterReconciler := &RayClusterReconciler{
+		Client:   fakeClient,
+		Recorder: &record.FakeRecorder{},
+		Scheme:   scheme.Scheme,
+		Log:      ctrl.Log.WithName("controllers").WithName("RayCluster"),
+	}
+
+	err = testRayClusterReconciler.reconcilePods(testRayCluster)
+	assert.Nil(t, err, "Fail to reconcile Pods")
+
+	err = fakeClient.List(context.Background(), &podList, &client.ListOptions{
+		LabelSelector: workerSelector,
+		Namespace:     namespaceStr,
+	})
+
+	assert.Nil(t, err, "Fail to get pod list after reconcile")
+
+	assert.Equal(t, int(expectReplicaNum), len(podList.Items),
+		"Replica number is wrong after reconcile expect %d actual %d", expectReplicaNum, len(podList.Items))
+}
+
+func TestReconcile_RandomDelete_OK(t *testing.T) {
+	setupTest(t)
+	defer tearDown(t)
+
+	var localExpectReplicaNum int32 = 2
+	testRayCluster.Spec.WorkerGroupSpecs[0].Replicas = &localExpectReplicaNum
+
+	fakeClient := clientFake.NewClientBuilder().WithRuntimeObjects(testPods...).Build()
+
+	podList := corev1.PodList{}
+	err := fakeClient.List(context.Background(), &podList, client.InNamespace(namespaceStr))
+
+	assert.Nil(t, err, "Fail to get pod list")
+
+	assert.Equal(t, len(testPods), len(podList.Items), "Init pod list len is wrong")
+
+	testRayClusterReconciler := &RayClusterReconciler{
+		Client:   fakeClient,
+		Recorder: &record.FakeRecorder{},
+		Scheme:   scheme.Scheme,
+		Log:      ctrl.Log.WithName("controllers").WithName("RayCluster"),
+	}
+
+	err = testRayClusterReconciler.reconcilePods(testRayCluster)
+	assert.Nil(t, err, "Fail to reconcile Pods")
+
+	err = fakeClient.List(context.Background(), &podList, &client.ListOptions{
+		LabelSelector: workerSelector,
+		Namespace:     namespaceStr,
+	})
+
+	assert.Nil(t, err, "Fail to get pod list after reconcile")
+
+	assert.Equal(t, int(localExpectReplicaNum), len(podList.Items),
+		"Replica number is wrong after reconcile expect %d actual %d", expectReplicaNum, len(podList.Items))
+
+	for i := 0; i < len(podList.Items); i++ {
+		if contains(testRayCluster.Spec.WorkerGroupSpecs[0].ScaleStrategy.WorkersToDelete, podList.Items[i].Name) {
```

Comment:
This part was not clear to me when reading the code -- I would expect the `WorkersToDelete` to be cleared already. But maybe that's only the case in the `PrioritizeWorkersToDelete = true` codepath :)

Reply:
(the same comment applies to the tests below)

Reply:
yes, will update the code